pkg/kt/cluster/kubernetes.go

```diff
@@ -69,30 +71,125 @@ func (k *Kubernetes) Deployment(name, namespace string) (*appv1.Deployment, erro
 	component, version := labels["kt-component"], labels["version"]
 	sshcm = fmt.Sprintf("kt-%s-public-key-%s", component, version)
 
-	generator, err := util.Generate(util.PrivateKeyPath(component, version))
+	privateKeyPath := util.PrivateKeyPath(component, version)
+
+	if reuseShadow {
+		pod, generator, err2 := k.tryGetExistingShadowRelatedObjs(name, namespace, sshcm, privateKeyPath, labels)
+		if err2 != nil {
+			err = err2
+			return
+		}
+		if pod != nil && generator != nil {
+			podIP, podName, credential = shadowResult(*pod, generator)
+			return
+		}
+	}
+
+	pod, generator, err := k.createShadowDeployment(labels, sshcm, namespace, privateKeyPath, name, image, debug)
 	if err != nil {
 		return
 	}
+	podIP, podName, credential = shadowResult(*pod, generator)
+	return
+}
+
+func (k *Kubernetes) tryGetExistingShadowRelatedObjs(name string, namespace string, sshcm string, privateKeyPath string, labels map[string]string) (pod *v1.Pod, generator *util.SSHGenerator, err error) {
+	_, shadowError := k.GetDeployment(name, namespace)
+	if shadowError != nil {
+		return
+	}
+	cli := k.Clientset.CoreV1().ConfigMaps(namespace)
+	configMap, configMapError := cli.Get(sshcm, metav1.GetOptions{})
+
+	if configMapError != nil {
+		err = errors.New("Found shadow deployment but no configMap. Please delete the deployment #{shadow}")
+		return
+	}
+
+	generator = util.NewSSHGenerator(configMap.Data[vars.SSHAuthPrivateKey], configMap.Data[vars.SSHAuthKey], privateKeyPath)
+
+	err = util.WritePrivateKey(generator.PrivateKeyPath, []byte(configMap.Data[vars.SSHAuthPrivateKey]))
+	if err != nil {
+		return
+	}
+
+	podList, err2 := k.Clientset.CoreV1().Pods(namespace).List(metav1.ListOptions{
+		LabelSelector: k8sLabels.Set(metav1.LabelSelector{MatchLabels: labels}.MatchLabels).String(),
+	})
+	if err2 != nil {
+		err = err2
+		return
+	}
+	if len(podList.Items) > 1 {
+		err = errors.New("Found more than one pod with name " + name + ", please make sure these is only one in namespace " + namespace)
+		return
+	}
+	if len(podList.Items) == 1 {
+		log.Info().Msgf("Found shared shadow, reuse it")
+		err = increaseRefCount(name, k.Clientset, namespace)
+		if err != nil {
+			return
+		}
+		return &(podList.Items[0]), generator, nil
+	}
+	err = errors.New("No Shadow pod found while shadow deployment present. You may need to clean up the deployment by yourself")
+	return
+}
+
+func increaseRefCount(name string, clientSet kubernetes.Interface, namespace string) error {
+	deployment, err := clientSet.AppsV1().Deployments(namespace).Get(name, metav1.GetOptions{})
+	if err != nil {
+		return err
+	}
+	label := deployment.ObjectMeta.Labels
+	count, err := strconv.Atoi(label[vars.RefCount])
+	if err != nil {
+		return err
+	}
+
+	deployment.ObjectMeta.Labels[vars.RefCount] = strconv.Itoa(count + 1)
+
+	_, err = clientSet.AppsV1().Deployments(namespace).Update(deployment)
```

Comment:
This operation seems like trigger k8s recreate-pod all online connection will be interrupt.